cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/RemoteAgentService.java

```diff
@@ -0,0 +1,165 @@
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpResponse;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * The main class to run the remote agent service.
+ */
+public class RemoteAgentService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteAgentService.class);
+  private static final Gson GSON = new Gson();
+  public static final String CONNECT_CONTROL_CHANNEL = "/v3/tethering/controlchannels/";
+
+  private final CConfiguration cConf;
+  private final TetherStore store;
+  private String instanceName;
+  private ScheduledExecutorService executorService;
+
+  @Inject
+  public RemoteAgentService(CConfiguration cConf, TransactionRunner transactionRunner) {
+    this.cConf = cConf;
+    this.store = new TetherStore(transactionRunner);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    instanceName = cConf.get(Constants.INSTANCE_NAME);
+    int connectionInterval = cConf.getInt(Constants.Tether.CONNECT_INTERVAL,
+                                          Constants.Tether.CONNECT_INTERVAL_DEFAULT);
+    executorService = Executors
+      .newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("tether-control-channel"));
+    executorService.scheduleWithFixedDelay(
+      () -> {
+        List<PeerInfo> peers = store.getPeers().stream()
+          // Ignore peers in REJECTED state.
+          .filter(p -> p.getTetherStatus() != TetherStatus.REJECTED)
+          .collect(Collectors.toList());
+        for (PeerInfo peer : peers) {
+          if (peer.getEndpoint() == null) {
+            // Should not happen
+            LOG.error("Peer {} does not have endpoint set", peer.getName());
+            continue;
+          }
+          try {
+            HttpResponse resp = TetherUtils.sendHttpRequest(HttpMethod.GET, new URI(peer.getEndpoint())
+              .resolve(CONNECT_CONTROL_CHANNEL + instanceName));
+            switch (resp.getResponseCode()) {
+              case 200:
+                TetherStatus peerStatus = store.getTetherStatus(peer.getName());
+                 if (peerStatus == TetherStatus.PENDING) {
+                  LOG.info("Peer {} transitioned to ACCEPTED state", peer.getName());
+                  store.updatePeer(peer.getName(), TetherStatus.ACCEPTED);
+                }
+                // Update last connection timestamp.
+                store.updatePeer(peer.getName(), System.currentTimeMillis());
```

Comment:
Is it possible to move this at store level? So that caller doesnt have to do it every time they call update?

Reply:
done.